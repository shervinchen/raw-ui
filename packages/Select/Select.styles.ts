import css from 'styled-jsx/css';
import { useTheme } from '../Theme';
import {
  SelectBasicStyles,
  SelectHoverStyles,
  SelectProps,
  SelectSizes,
  SelectSizeStyles,
  SelectStyles,
  SelectTypes,
} from './Select.types';

export const useSelectStyles = ({
  type,
  size,
  disabled,
}: SelectProps): SelectStyles => {
  const theme = useTheme();

  const sizes: {
    [key in SelectSizes]: SelectSizeStyles;
  } = {
    xs: {
      fontSize: '12px',
      height: '24px',
      paddingLeft: '8px',
      paddingRight: '32px',
      iconRight: '8px',
      tagContentVerticalPadding: '2px',
      tagHeight: '18px',
    },
    sm: {
      fontSize: '14px',
      height: '28px',
      paddingLeft: '8px',
      paddingRight: '32px',
      iconRight: '8px',
      tagContentVerticalPadding: '2px',
      tagHeight: '22px',
    },
    md: {
      fontSize: '14px',
      height: '32px',
      paddingLeft: '12px',
      paddingRight: '40px',
      iconRight: '12px',
      tagContentVerticalPadding: '2px',
      tagHeight: '26px',
    },
    lg: {
      fontSize: '14px',
      height: '36px',
      paddingLeft: '12px',
      paddingRight: '40px',
      iconRight: '12px',
      tagContentVerticalPadding: '2px',
      tagHeight: '30px',
    },
    xl: {
      fontSize: '16px',
      height: '40px',
      paddingLeft: '12px',
      paddingRight: '40px',
      iconRight: '12px',
      tagContentVerticalPadding: '2px',
      tagHeight: '34px',
    },
  };

  const styles: {
    [key in SelectTypes]: SelectBasicStyles;
  } = {
    default: {
      color: theme.palette.foreground,
      borderColor: theme.palette.accents2,
    },
    warning: {
      color: theme.palette.foreground,
      borderColor: theme.palette.warning,
    },
    error: {
      color: theme.palette.foreground,
      borderColor: theme.palette.error,
    },
  };

  const defaultStyles = {
    ...(styles?.[type || 'default'] ?? styles['default']),
    ...(sizes?.[size || 'md'] ?? sizes['md']),
  };

  const disabledStyles: SelectBasicStyles = {
    backgroundColor: theme.palette.accents1,
    borderColor: theme.palette.accents2,
    color: theme.palette.accents6,
    cursor: 'not-allowed',
  };

  return {
    ...defaultStyles,
    ...(disabled ? disabledStyles : {}),
  };
};

export const useSelectHoverStyles = ({
  type,
  disabled,
}: SelectProps): SelectHoverStyles => {
  const theme = useTheme();

  const styles: {
    [key in SelectTypes]: SelectHoverStyles;
  } = {
    default: {
      hoverBorderColor: theme.palette.accents7,
    },
    warning: {
      hoverBorderColor: theme.palette.warning,
    },
    error: {
      hoverBorderColor: theme.palette.error,
    },
  };

  const defaultStyles = {
    ...(styles?.[type || 'default'] ?? styles['default']),
  };

  return disabled ? {} : defaultStyles;
};

export const useSelectCSS = ({ width, type, size, disabled }: SelectProps) => {
  const theme = useTheme();
  const {
    fontSize,
    height,
    paddingLeft,
    paddingRight,
    iconRight,
    tagContentVerticalPadding,
    color,
    borderColor,
    backgroundColor = theme.palette.background,
    cursor = 'pointer',
  } = useSelectStyles({ type, size, disabled });
  const { hoverBorderColor = borderColor } = useSelectHoverStyles({
    type,
    disabled,
  });

  return css.resolve`
    .raw-select {
      box-sizing: border-box;
      display: inline-flex;
      position: relative;
      align-items: center;
      width: ${width};
      height: ${height};
      padding-left: ${paddingLeft};
      padding-right: ${paddingRight};
      border: 1px solid ${borderColor};
      border-radius: 6px;
      background-color: ${backgroundColor};
      transition: border-color 0.15s ease, color 0.15s ease;
      cursor: ${cursor};
      user-select: none;
    }
    .raw-select.multiple {
      height: auto;
      min-height: ${height};
    }
    .raw-select :global(.raw-select-inner) {
      display: inline-flex;
      width: 100%;
    }
    .raw-select :global(.raw-select-placeholder) {
      max-width: 100%;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
      font-size: ${fontSize};
      color: ${theme.palette.accents5};
    }
    .raw-select :global(.raw-select-content) {
      max-width: 100%;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
      font-size: ${fontSize};
      color: ${color};
    }
    .raw-select :global(.raw-select-tag-content) {
      display: flex;
      flex-wrap: wrap;
      padding-top: ${tagContentVerticalPadding};
      padding-bottom: ${tagContentVerticalPadding};
      margin: -2px;
    }
    .raw-select :global(.raw-select-arrow) {
      display: inline-flex;
      align-items: center;
      position: absolute;
      right: ${iconRight};
      top: 50%;
      transform: translateY(-50%);
      pointer-events: none;
      color: ${theme.palette.accents7};
    }
    .raw-select :global(.raw-select-clear) {
      display: inline-flex;
      align-items: center;
      position: absolute;
      right: ${iconRight};
      top: 50%;
      transform: translateY(-50%);
      color: ${theme.palette.foreground};
    }
    .raw-select.raw-select-active,
    .raw-select:hover {
      border-color: ${hoverBorderColor};
    }
    .raw-select.raw-select-active :global(.raw-select-arrow),
    .raw-select:hover :global(.raw-select-arrow) {
      color: ${disabled ? theme.palette.accents7 : theme.palette.foreground};
    }
  `;
};
