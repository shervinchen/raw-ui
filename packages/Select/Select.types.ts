import { HTMLAttributes, MutableRefObject } from 'react';

export type SelectOptionValue = string | number | undefined | null;

export type SelectValue =
  | SelectOptionValue
  | SelectOptionValue[]
  | undefined
  | null;

export type SelectTypes = 'default' | 'warning' | 'error';

export type SelectSize = 'sm' | 'md' | 'lg';

export interface BaseSelectProps {
  value?: SelectValue;
  defaultValue?: SelectValue;
  width?: string;
  placeholder?: string;
  type?: SelectTypes;
  size?: SelectSize;
  disabled?: boolean;
  multiple?: boolean;
  clearable?: boolean;
  className?: string;
  dropdownClassName?: string;
  dropdownHeight?: string;
  getPopupContainer?: () => HTMLElement | null;
  onChange?: (value: SelectValue) => void;
}

type NativeSelectProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseSelectProps
>;

export type SelectProps = BaseSelectProps & NativeSelectProps;

export type SelectRef = {
  focus: () => void;
  blur: () => void;
};

export interface SelectConfig {
  multiple?: boolean;
  selectValue?: SelectValue;
  onSelectChange?: (optionValue?: SelectOptionValue) => void;
  selectRef?: MutableRefObject<HTMLElement | null>;
  getPopupContainer?: () => HTMLElement | null;
  dropdownHeight?: string;
  selectDisabled?: boolean;
  selectId: string;
}

export interface SelectSizeStyles {
  fontSize?: string;
  height?: string;
  horizontalPadding?: string;
}

export interface SelectBasicStyles {
  borderColor?: string;
  color?: string;
  backgroundColor?: string;
  cursor?: 'pointer' | 'not-allowed';
}

export type SelectStyles = SelectSizeStyles & SelectBasicStyles;

export interface SelectHoverStyles {
  hoverBorderColor?: string;
}
