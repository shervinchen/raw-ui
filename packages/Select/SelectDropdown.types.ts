import { ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

export interface BaseSelectDropdownProps {
  visible: boolean;
  className?: string;
}

export type SelectDropdownProps = Merge<
  ComponentPropsWithoutRef<'div'>,
  BaseSelectDropdownProps
>;
