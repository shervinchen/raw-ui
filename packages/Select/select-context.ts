import { createContext, useContext } from 'react';
import { SelectConfig } from './Select.types';

const defaultContext: SelectConfig = {
  selectId: '',
  multiple: false,
  disabled: false,
};

export const SelectContext = createContext<SelectConfig>(defaultContext);

export const useSelectContext = (): SelectConfig =>
  useContext<SelectConfig>(SelectContext);
