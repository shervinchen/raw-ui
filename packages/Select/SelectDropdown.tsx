import React, {
  forwardRef,
  PropsWithChildren,
  useRef,
  useImperativeHandle,
} from 'react';
import classNames from 'classnames';
import { SelectDropdownProps } from './SelectDropdown.types';
import { useTransition } from '../utils/hooks';
import { RawUITheme } from '../Theme/preset/preset.type';
import { useTheme } from '../Theme/theme-context';
import { useSelectContext } from './select-context';
import Popup from '../Popup';
import { computeTargetPosition } from '../Popup/computePopup';

const SelectDropdown = forwardRef<
  HTMLDivElement | null,
  PropsWithChildren<SelectDropdownProps>
>(({ visible, className, children }, ref) => {
  const theme: RawUITheme = useTheme();
  const dropdownRef = useRef<HTMLDivElement | null>(null);
  const { multiple, dropdownHeight, selectRef, getPopupContainer, selectId } =
    useSelectContext();
  const { stage: dropdownTransitionStage } = useTransition(visible, 0, 0);
  const dropdownClasses = classNames('raw-select-dropdown', className);
  const selectRect = selectRef?.current?.getBoundingClientRect() ?? null;
  const dropdownWidth = selectRect
    ? selectRect.width || selectRect.right - selectRect.left
    : 0;

  useImperativeHandle<HTMLDivElement | null, HTMLDivElement | null>(
    ref,
    () => dropdownRef.current
  );

  return (
    <Popup
      name="dropdown"
      visible={visible}
      targetRef={selectRef}
      getPopupPosition={() => {
        const { bottom, left } = computeTargetPosition(
          selectRef,
          getPopupContainer
        );
        return {
          top: bottom,
          left,
          transform: 'translate(0, 0)',
        };
      }}
      getPopupContainer={getPopupContainer}
    >
      <div
        data-testid="selectDropdown"
        role="listbox"
        aria-multiselectable={multiple}
        id={selectId}
        ref={dropdownRef}
        className={dropdownClasses}
        style={{
          opacity: dropdownTransitionStage === 'enter' ? 1 : 0,
        }}
      >
        {children}
        <style jsx>{`
          .raw-select-dropdown {
            width: ${dropdownWidth}px;
            padding: 8px;
            margin-top: 2px;
            border-radius: 6px;
            border: 1px solid ${theme.palette.accents2};
            box-shadow: ${theme.tokens.shadow.sm};
            background-color: ${theme.palette.background};
            max-height: ${dropdownHeight};
            overflow-y: auto;
            transition: opacity 0.15s ease;
          }
        `}</style>
      </div>
    </Popup>
  );
});

SelectDropdown.displayName = 'RawSelectDropdown';

export default SelectDropdown;
