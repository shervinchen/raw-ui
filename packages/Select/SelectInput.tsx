import React, {
  ComponentPropsWithRef,
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef,
} from 'react';
import { VisuallyHiddenInput } from '../VisuallyHidden';
import { useSelectContext } from './select-context';

type SelectInputProps = {
  visible: boolean;
  onBlur: () => void;
  onFocus: () => void;
};

const SelectInput = forwardRef(
  (
    { visible, onBlur, onFocus }: SelectInputProps,
    inputRef: ComponentPropsWithRef<'input'>['ref']
  ) => {
    const ref = useRef<HTMLInputElement | null>(null);
    const { selectId } = useSelectContext();

    useImperativeHandle(inputRef, () => ref.current);

    useEffect(() => {
      if (visible) {
        ref.current?.focus({
          preventScroll: true,
        });
      }
    }, [visible]);

    return (
      <>
        <VisuallyHiddenInput
          role="combobox"
          aria-haspopup="listbox"
          aria-expanded={visible}
          aria-controls={selectId}
          ref={ref}
          type="search"
          autoComplete="off"
          readOnly
          unselectable="on"
          onBlur={onBlur}
          onFocus={onFocus}
        />
      </>
    );
  }
);

SelectInput.displayName = 'RawSelectInput';

export default SelectInput;
