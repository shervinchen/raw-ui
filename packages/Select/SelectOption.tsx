import React, { FC, PropsWithChildren, MouseEvent, useMemo } from 'react';
import { Check } from 'react-feather';
import classNames from 'classnames';
import { SelectOptionProps } from './SelectOption.types';
import { RawUITheme } from '../Theme/preset/preset.type';
import { useTheme } from '../Theme/theme-context';
import { useSelectContext } from './select-context';
import { useSelectStyles } from './Select.styles';

const SelectOption: FC<PropsWithChildren<SelectOptionProps>> = ({
  value,
  disabled = false,
  className = '',
  children,
  ...restProps
}) => {
  const theme: RawUITheme = useTheme();
  const {
    multiple,
    selectValue,
    onSelectChange,
    type,
    size,
    disabled: selectDisabled,
  } = useSelectContext();
  const { fontSize, height, paddingLeft } = useSelectStyles({
    type,
    size,
    disabled: selectDisabled,
  });
  const classes = classNames('raw-select-option', className);
  const isDisabled = selectDisabled || disabled;
  const isSelected = useMemo(() => {
    if (Array.isArray(selectValue)) {
      return selectValue.includes(value);
    } else {
      return multiple ? false : selectValue === value;
    }
  }, [multiple, selectValue, value]);

  const handleClick = (event: MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    if (isDisabled) return;
    onSelectChange?.(value);
  };

  return (
    <div
      data-testid="selectOption"
      role="option"
      aria-selected={isSelected}
      aria-disabled={isDisabled}
      className={classes}
      {...restProps}
      onClick={handleClick}
    >
      <div className="raw-select-option-content">
        <span className="raw-select-option-text">{children}</span>
        {isSelected && (
          <div className="raw-select-option-check">
            <Check size={16} color={theme.palette.neutral['600']} />
          </div>
        )}
      </div>
      <style jsx>{`
        .raw-select-option {
          box-sizing: border-box;
          display: flex;
          align-items: center;
          width: 100%;
          height: ${height};
          padding: 0 ${paddingLeft};
          font-size: ${fontSize};
          color: ${isDisabled
            ? theme.palette.neutral['500']
            : theme.palette.neutral['600']};
          background-color: ${theme.palette.background};
          border-radius: 6px;
          transition: background-color 0.15s ease;
          cursor: ${isDisabled ? 'not-allowed' : 'pointer'};
          user-select: none;
        }
        .raw-select-option:hover {
          background-color: ${isDisabled
            ? theme.palette.background
            : theme.palette.neutral['200']};
        }
        .raw-select-option-content {
          display: flex;
          align-items: center;
          width: 100%;
        }
        .raw-select-option-text {
          overflow: hidden;
          white-space: nowrap;
          text-overflow: ellipsis;
        }
        .raw-select-option-check {
          display: inline-flex;
          align-items: center;
          margin-left: auto;
          flex-shrink: 0;
        }
      `}</style>
    </div>
  );
};

export default SelectOption;
