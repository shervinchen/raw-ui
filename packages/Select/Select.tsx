import React, {
  forwardRef,
  PropsWithChildren,
  MouseEvent,
  useMemo,
  useState,
  useRef,
  useImperativeHandle,
  ReactElement,
  useCallback,
  useId,
} from 'react';
import { ChevronDown } from 'react-feather';
import classNames from 'classnames';
import { useClickAway } from 'react-use';
import {
  SelectConfig,
  SelectOptionValue,
  SelectProps,
  SelectRef,
  SelectValue,
} from './Select.types';
import { useControlled } from '../utils/hooks';
import { getValidChildren } from '../utils/common';
import { SelectContext } from './select-context';
import SelectDropdown from './SelectDropdown';
import SelectInput from './SelectInput';
import SelectTag from './SelectTag';
import { useSelectCSS } from './Select.styles';

const getInternalValue = (multiple: boolean, value: SelectValue) => {
  if (Array.isArray(value)) {
    return multiple ? value : undefined;
  } else {
    return multiple ? undefined : value;
  }
};

const getNewInternalValue = (
  multiple: boolean,
  prevValue: SelectValue,
  nextValue: SelectOptionValue
) => {
  if (multiple) {
    if (Array.isArray(prevValue)) {
      const isExist = prevValue.includes(nextValue);
      return isExist
        ? prevValue.filter((item) => item !== nextValue)
        : [...prevValue, nextValue];
    } else {
      return [nextValue];
    }
  } else {
    return nextValue;
  }
};

const sortSelectedOptions = (
  selectedValue: SelectOptionValue[],
  selectedOption: ReactElement[]
): ReactElement[] => {
  return selectedOption.sort(
    (a, b) =>
      selectedValue.indexOf(a.props.value) -
      selectedValue.indexOf(b.props.value)
  );
};

const Select = forwardRef<SelectRef, PropsWithChildren<SelectProps>>(
  (
    {
      defaultValue,
      value,
      width = '100%',
      type = 'default',
      // size = "md",
      disabled = false,
      multiple = false,
      // clearable = false,
      placeholder = 'Select option',
      className = '',
      dropdownClassName = '',
      dropdownHeight = 'none',
      getPopupContainer,
      onChange,
      children,
      ...restProps
    },
    ref
  ) => {
    const selectId = `raw-select-dropdown-${useId()}`;
    const selectRef = useRef<HTMLDivElement>(null);
    const inputRef = useRef<HTMLInputElement>(null);
    const dropdownRef = useRef<HTMLDivElement>(null);
    const [internalValue, setInternalValue] = useControlled<SelectValue>({
      defaultValue: getInternalValue(multiple, defaultValue),
      value: getInternalValue(multiple, value),
    });
    const [selectFocus, setSelectFocus] = useState(false);
    const [dropdownVisible, setDropdownVisible] = useState(false);
    const { className: resolveClassName, styles } = useSelectCSS({
      width,
      type,
      disabled,
      dropdownVisible,
    });
    const selectClasses = classNames(
      'raw-select',
      (selectFocus || dropdownVisible) && 'raw-select-active',
      multiple && 'multiple',
      className,
      resolveClassName
    );

    const clickHandler = (event: MouseEvent<HTMLDivElement>) => {
      event.stopPropagation();
      if (disabled) return;
      setDropdownVisible(!dropdownVisible);
    };

    const mouseDownHandler = (event: MouseEvent<HTMLDivElement>) => {
      event.preventDefault();
    };

    const changeHandler = useCallback(
      (optionValue?: SelectOptionValue) => {
        if (!multiple) {
          setDropdownVisible(false);
        }
        const newInternalValue = getNewInternalValue(
          multiple,
          internalValue,
          optionValue
        );
        setInternalValue(newInternalValue);
        onChange?.(newInternalValue);
      },
      [internalValue, multiple, onChange, setInternalValue]
    );

    const focusHandler = () => setSelectFocus(true);

    const blurHandler = () => setSelectFocus(false);

    const selectConfig = useMemo<SelectConfig>(() => {
      return {
        multiple,
        selectValue: internalValue,
        onSelectChange: changeHandler,
        selectRef,
        dropdownHeight,
        getPopupContainer,
        selectDisabled: disabled,
        selectId,
      };
    }, [
      multiple,
      internalValue,
      changeHandler,
      dropdownHeight,
      getPopupContainer,
      disabled,
      selectId,
    ]);

    useClickAway(
      selectRef,
      () => {
        setDropdownVisible(false);
      },
      ['mousedown']
    );

    useImperativeHandle(
      ref,
      () => ({
        focus: () => inputRef.current?.focus(),
        blur: () => inputRef.current?.blur(),
      }),
      [inputRef]
    );

    const SelectContent = (): ReactElement => {
      if (internalValue === undefined) {
        return <span className="raw-select-placeholder">{placeholder}</span>;
      }

      const selectedOptions = getValidChildren(children).filter((option) => {
        if (Array.isArray(internalValue)) {
          return internalValue.includes(option.props.value);
        } else {
          return internalValue === option.props.value;
        }
      });

      const isEmptyValue =
        (!Array.isArray(internalValue) && internalValue === '') ||
        (Array.isArray(internalValue) && internalValue.length === 0);

      if (selectedOptions.length === 0) {
        if (isEmptyValue) {
          return <span className="raw-select-placeholder">{placeholder}</span>;
        } else {
          return <span className="raw-select-content"></span>;
        }
      }

      if (multiple) {
        return (
          <div className="raw-select-tag-content">
            {sortSelectedOptions(
              internalValue as SelectOptionValue[],
              selectedOptions
            ).map((option) => (
              <SelectTag
                key={option.props.value as SelectOptionValue}
                disabled={disabled}
                onDeleteTag={() => {
                  changeHandler(option.props.value);
                }}
              >
                {option.props.children}
              </SelectTag>
            ))}
          </div>
        );
      } else {
        return (
          <span className="raw-select-content">
            {selectedOptions[0].props.children}
          </span>
        );
      }
    };

    return (
      <SelectContext.Provider value={selectConfig}>
        <div
          data-testid="selectContainer"
          className={selectClasses}
          ref={selectRef}
          onClick={clickHandler}
          onMouseDown={mouseDownHandler}
          {...restProps}
        >
          <div className="raw-select-inner">
            <SelectContent />
            <SelectInput
              ref={inputRef}
              visible={dropdownVisible}
              onBlur={blurHandler}
              onFocus={focusHandler}
            />
          </div>
          <div className="raw-select-arrow">
            <ChevronDown size={16} />
          </div>
          <SelectDropdown
            ref={dropdownRef}
            visible={dropdownVisible}
            className={dropdownClassName}
          >
            {children}
          </SelectDropdown>
          {styles}
        </div>
      </SelectContext.Provider>
    );
  }
);

Select.displayName = 'RawSelect';

export default Select;
