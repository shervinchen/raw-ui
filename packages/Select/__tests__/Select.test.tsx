import React, { useState } from 'react';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Select from '..';
import { SelectProps, SelectValue } from '../Select.types';

const optionsData = [
  {
    name: 'React',
    value: 'react',
  },
  {
    name: 'Vue',
    value: 'vue',
  },
  {
    name: 'Angular',
    value: 'angular',
  },
  {
    name: 'Svelte',
    value: 'svelte',
  },
];

describe('Select', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Select>
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should forward ref', () => {
    const ref = React.createRef<HTMLInputElement>();
    render(
      <Select ref={ref}>
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    const selectInput = screen.getByTestId('selectInput');
    ref.current.focus();
    expect(selectInput).toHaveFocus();
    ref.current.blur();
    expect(selectInput).not.toHaveFocus();
  });

  test('should support custom class name', () => {
    const { container } = render(
      <Select className="custom-select">
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    expect(container.firstChild).toHaveClass('custom-select');
  });

  test('should support placeholder', () => {
    render(
      <Select placeholder="Select option">
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    expect(screen.getByText('Select option')).toBeInTheDocument();
  });

  test('should support custom width', () => {
    const { container } = render(
      <Select width="200px">
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    const select = container.firstChild as Element;
    expect(getComputedStyle(select).width).toBe('200px');
  });

  test('should support uncontrolled value', () => {
    render(
      <Select defaultValue="1">
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    expect(document.querySelector('.raw-select').innerHTML).toContain(
      'Option 1'
    );
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: SelectProps) => {
      const [selectValue, setSelectValue] = useState<SelectValue>();

      return (
        <Select
          value={selectValue}
          onChange={(value) => {
            setSelectValue(value);
            props.onChange?.(value);
          }}
        >
          <Select.Option value="1">Option 1</Select.Option>
          <Select.Option value="2">Option 2</Select.Option>
        </Select>
      );
    };
    const { container } = render(<Component onChange={onChange} />);
    const select = container.firstChild as Element;
    const option1 = document.querySelectorAll('.raw-select-option')[0];
    await userEvent.click(select);
    await userEvent.click(option1);
    setTimeout(() => {
      expect(document.querySelector('.raw-select').innerHTML).toContain(
        'Option 1'
      );
    }, 150);
  });

  test('should support multiple value', () => {
    render(
      <Select multiple defaultValue={['react', 'vue']}>
        {optionsData.map((item) => (
          <Select.Option value={item.value} key={item.value}>
            {item.name}
          </Select.Option>
        ))}
      </Select>
    );
    expect(document.querySelector('.raw-select').innerHTML).toContain('React');
    expect(document.querySelector('.raw-select').innerHTML).toContain('Vue');
  });

  test('should support select disabled', async () => {
    const { container } = render(
      <Select disabled>
        <Select.Option value="1">Option 1</Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    const select = container.firstChild as Element;
    await userEvent.click(select);
    expect(document.querySelectorAll('.raw-select-option').length).toBe(0);
  });

  test('should support option disabled', async () => {
    const { container } = render(
      <Select>
        <Select.Option value="1" disabled>
          Option 1
        </Select.Option>
        <Select.Option value="2">Option 2</Select.Option>
      </Select>
    );
    const select = container.firstChild as Element;
    const option1 = document.querySelectorAll('.raw-select-option')[0];
    const option2 = document.querySelectorAll('.raw-select-option')[1];
    await userEvent.click(select);
    await userEvent.click(option1);
    expect(document.querySelector('.raw-select').innerHTML).not.toContain(
      'Option 1'
    );
    await userEvent.click(option2);
    setTimeout(() => {
      expect(document.querySelector('.raw-select').innerHTML).toContain(
        'Option 2'
      );
    }, 150);
  });
});
