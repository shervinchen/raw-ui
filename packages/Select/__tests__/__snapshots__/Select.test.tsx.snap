// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="raw-select jsx-1520639952"
    data-testid="selectContainer"
  >
    <div
      class="raw-select-inner"
    >
      <span
        class="raw-select-placeholder"
      >
        Select option
      </span>
      <input
        aria-controls="raw-select-dropdown-:r0:"
        aria-expanded="false"
        aria-haspopup="listbox"
        autocomplete="off"
        class="jsx-1802955700 "
        readonly=""
        role="combobox"
        type="search"
        unselectable="on"
      />
    </div>
    <div
      class="raw-select-arrow"
    >
      <svg
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
